simplified-ui-catalog/src/main/java/org/nypl/simplified/ui/catalog/CatalogFragmentLoginDialog.kt

```diff
@@ -328,29 +306,63 @@ class CatalogFragmentLoginDialog : DialogFragment() {
 
       is AccountLoginState.AccountLoggingIn -> {
         this.lockForm()
+        this.errorDetails.visibility = View.GONE
         this.progress.visibility = View.VISIBLE
         this.progressText.text = state.status
         this.progressText.visibility = View.VISIBLE
       }
 
       is AccountLoginState.AccountLoginFailed -> {
         this.unlockForm()
+        this.errorDetails.visibility = View.VISIBLE
         this.action.isEnabled = this.determineLoginIsSatisfied()
         this.progress.visibility = View.INVISIBLE
         this.progressText.visibility = View.VISIBLE
+        this.progressText.text = state.taskResult.steps.last().resolution.message
         this.action.setOnClickListener {
           this.lockForm()
           this.tryLogin()
         }
+        this.errorDetails.setOnClickListener {
+          this.openErrorPage(state.taskResult.steps)
+        }
+      }
+
+      is AccountLoginState.AccountLogoutFailed -> {
+        this.lockForm()
+        this.errorDetails.visibility = View.VISIBLE
+        this.action.isEnabled = this.determineLoginIsSatisfied()
+        this.progress.visibility = View.INVISIBLE
+        this.progressText.visibility = View.VISIBLE
+        this.progressText.text = state.taskResult.steps.last().resolution.message
+        this.errorDetails.setOnClickListener {
+          this.openErrorPage(state.taskResult.steps)
+        }
       }
 
       is AccountLoginState.AccountLoggedIn,
-      is AccountLoginState.AccountLoggingOut,
-      is AccountLoginState.AccountLogoutFailed ->
-        this.dismiss()
+      is AccountLoginState.AccountLoggingOut -> {
+        this.findNavigationController().popBackStack()
+        Unit
```

Comment:
If you remove the return at the top of this function you probably don't need Unit here.

Reply:
Unfortunately, that means that the `when` becomes a statement rather than expression and so isn't subjected to exhaustiveness checks. Later on, someone will add a new case to the `AccountLoginState` sealed class, and the compiler won't tell us that this `when` needs to be updated...

Reply:
Wow, I was actually not aware that the behavior of `when` changed when used as an expression. That's an interesting gotcha.

Reply:
I think there's a combination of problems here. The JVM doesn't yet have support for sealed classes (that's one of the things currently underway in Project Amber). The Kotlin designers decided that `when` statements need not be exhaustive, presumably for making code convenient to write :-1: 

Unfortunately, it doesn't make sense for a `when` _expression_ not to be exhaustive... In an expression context, every branch of the `when` _needs_ to either yield a value or not terminate. This implies that you need to handle every case. This makes expression `when` behave differently to statement `when`. Modern versions of Java have the same distinction... The old style statement `switch` doesn't need to be exhaustive, while a `switch` expression does.